---
title: "s100a8 Power Simulations"
author: "Daniel Gatti"
date: "January 20, 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(knitr)
library(pcaMethods)
library(readxl)
library(tidyverse)
library(qtl2)
base_dir = '/media/dmgatti/hdb/projects/TB'
pheno_dir = file.path(base_dir, 'data', 'phenotypes')
haplo_dir = file.path(base_dir, 'haplo_reconstr')
results_dir = file.path(base_dir, 'results', 'power_sim')
```

## Introduction

The goals off these simulations is to estimate the power to detect a QTL of a given size using different subset of mice.  I will sample 10 subsets of 100, 200 and 300 samples. There are a total of 408 samples. I will map only on Chr 3 and look at the s100a8 locus

Read in the data.

```{r read_data}
load(file.path(pheno_dir, 'GB_Tufts_qtl2_input.Rdata'))
```

Read in new phenotypes and haplotypes.

```{r rankz}
rankZ = function(x) {
  x = rank(x, na.last = "keep", ties.method = "average") / (sum(!is.na(x)) + 1)
  return(qnorm(x))
}
```


```{r}
new_pheno = read_xlsx(file.path(pheno_dir, 'Lung blood BW data GB07 GB07.1 GB07.2 GB07.3 GB07.4 GB07.5 GB07.6.xlsx'), 
                      sheet = 'Class, Morbidity, Lung, Serum')
new_pheno = new_pheno %>%
              filter(`Mouse strain` == 'J:DO' & substr(`GigaMUGA genotyped`, 1, 1) == 'y') %>%
              select(sample = `Mouse #`,
                     strain = `Mouse strain`,
                     batch  = `Experiment (Batch)`, 
                     gm_geno = `GigaMUGA genotyped`, 
                     dose    = `Mtb initial dose`,
                     wt_loss = `% wt loss`,
                     mtb_burden = `Lung Mtb burden`,
                     cxcl5  = `Lung CXCL5 pg/ml`,
                     cxcl2  = `Lung CXCL2 pg/ml`,
                     cxcl1  = `Lung CXCL1 pg/ml`,
                     ifng   = `Lung IFNg pg/ml`,
                     tnf    = `Lung TNF pg/ml`,
                     il12   = `Lung IL-12 pg/ml`,
                     il10   = `Lung IL-10 pg/ml`,
                     mmp8   = `Lung MMP8 pg/ml`,
                     vegf   = `Lung VEGF pg/ml`,
                     s100a8 = `Lung S100A8 pg/ml`) %>%
              mutate(wt_loss    = rankZ(as.numeric(wt_loss)),
                     mtb_burden = rankZ(as.numeric(mtb_burden)),
                     cxcl5      = rankZ(as.numeric(cxcl5)),
                     cxcl2      = rankZ(as.numeric(cxcl2)),
                     cxcl1      = rankZ(as.numeric(cxcl1)),
                     ifng       = rankZ(as.numeric(ifng)),
                     tnf        = rankZ(as.numeric(tnf)),
                     il12      = rankZ(as.numeric(il12)),
                     il10      = rankZ(as.numeric(il10)),
                     mmp8      = rankZ(as.numeric(mmp8)),
                     vegf      = rankZ(as.numeric(vegf)),
                     s100a8    = rankZ(as.numeric(s100a8)))
pheno_rz = as.matrix(new_pheno[,6:ncol(new_pheno)])
rownames(pheno_rz) = new_pheno$sample

covar = model.matrix(~dose, data = new_pheno)[,-1]
rownames(covar) = rownames(pheno_rz)
```

```{r read_haplo}
new_probs = readRDS(file.path(haplo_dir, 'tufts_do_alleleprobs.rds'))

for(i in seq_along(new_probs)) {
  rownames(new_probs[[i]]) = sub('^g[0-9]_Tufts_GB(2)?_', '', rownames(new_probs[[i]]))
  rownames(new_probs[[i]]) = sub('^g[0-9]_', '', rownames(new_probs[[i]]))
} # for(i)
```

Intersect sample IDs.

```{r subste samples}
samples = intersect(rownames(pheno_rz), rownames(new_probs[[1]]))
pheno_rz = pheno_rz[samples,]
covar = covar[samples,]

for(i in seq_along(new_probs)) {
  new_probs[[i]] = new_probs[[i]][samples,,]
}
stopifnot(rownames(pheno_rz) == rownames(new_probs[[1]]))
probs = new_probs
rm(new_probs)
```

Make new kinship matrix.

```{r new _kinship}
K = calc_kinship(probs = probs, type = 'loco', cores = 4)
```

Estimate heritability of phenotypes.

```{r}
allK = calc_kinship(probs, cores = 4)
h2 = qtl2::est_herit(pheno_rz, kinship = allK, addcovar = covar)
write_csv(data.frame(Phenotype = names(h2), h2 = h2), path = file.path(results_dir, 'heritability.csv'))
kable(h2)
```

Heritability for s100a8 is 0.55.

```{r map_full}
qtl_full = scan1(genoprobs = probs, pheno = pheno_rz, kinship = K, addcovar = covar, cores = 4)
```

```{r}
for(i in colnames(qtl_full)) {
  plot_scan1(qtl_full, map, lodcolumn = i, main = i)
}
```

```{r}
out_dir = file.path(base_dir, 'results', 'qtl2', 'dose_as_factor_run2')
for(pheno_name in colnames(qtl_full)) {
  png(file.path(out_dir, paste0(pheno_name, '_qtl.png')), width = 2000, height = 1800, res = 300)
  plot_scan1(qtl_full, map, lodcolumn = pheno_name, main = pheno_name)
  add_threshold(map, thresholdA = 7.8, col = 2)
  dev.off()
}
```

```{r}
brks = cut(1:nrow(qtl_full), 600)
spl = split(data.frame(qtl_full), brks)
spl = lapply(spl, as.matrix)
spl = lapply(spl, apply, 2, max)
lod = matrix(unlist(spl), nrow = ncol(qtl_full), dimnames = list(colnames(qtl_full), 1:length(spl)))
heatmap(lod, Colv = NA)
```

Selecting Mtb Burden, Cxcl1, s100a8, and Tnf as phenotypes becuase they have a range of heritabilities. I will only need to map on Chr 3, 5 and 15.

| Phenotype  | Chr | Pos (Mb) |  LOD  |  h2  |
|:----------:|:---:|:--------:|:-----:|:----:|
| Mtb Burden | 15  |  25.751  |  6.98 | 0.54 |
| Cxcl1      | 15  |  25.751  |  9.09 | 0.61 |
| s100a8     |  3  |  90.687  | 15.09 | 0.56 |
| Tnf        |  1  | 154.232  |  7.57 | 0.35 |

Map using all of the samples, but only the genoprobs on Chr 1, 3, & 15.

```{r peaks_full_qtl}
#pheno_names = c('mtb_burden', 'cxcl1', 's100a8', 'tnf')
#chr_names = c('1', '3', '15')

pheno_names = c('mtb_burden', 'cxcl1', 'cxcl2', 'cxcl5', 's100a8', 'tnf')
chr_names = '15'

peaks_full = find_peaks(qtl_full, map = map, threshold = 6, prob = 0.95)
write.csv(peaks_full, file = file.path(base_dir, 'results', 'qtl2', 'dose_as_factor_run2', 'tufts_do_tb_qtl_peaks_525samples.csv'))


#peaks_full = subset(peaks_full, lodcolumn == 'mtb_burden' & chr == '15'  |
#                                lodcolumn == 'cxcl1'      & chr == '15' |
#                                lodcolumn == 's100a8'     & chr == '3'  |
#                                lodcolumn == 'tnf'        & chr == '1')

peaks_full = subset(peaks_full, lodcolumn %in% pheno_names & chr == '15')

peaks_full$marker = find_marker(map = map, chr = peaks_full$chr, pos = peaks_full$pos)
peaks_full$chr = as.character(peaks_full$chr)
kable(peaks_full)
```

For each number of samples, select a subset, map on Chr 3 and record the LOD.  The threshold changes with sample size...

```{r}
sample_size = c(100, 200, 300, 400, 500)
num_sims    = 20

result = data.frame(sim = NA, phenotype = NA, samples = NA, lod = NA, pos = NA, prox = NA, dist = NA)

sim = 1
for(ss in sample_size) {

  for(i in 1:num_sims) {
    
    print(paste('sim', sim, 'sample', ss))
    
    samples = sample(rownames(pheno_rz), ss)
    pr = probs[samples, chr_names]
    ph = pheno_rz[samples, pheno_names, drop = FALSE]
    k = K[chr_names]
    for(chr in seq_along(names(k))) {
       k[[chr]] = K[[chr]][samples, samples]
    }
    cv = covar[samples,]
    
    lod = scan1(genoprobs = pr, pheno = ph, kinship = k, addcovar = cv, cores = 4)

    pk = find_peaks(lod, map, threshold = 1, prob = 0.95)
    pk$chr = as.character(pk$chr)
#    pk = subset(pk, lodcolumn == peaks_full$lodcolumn[1] & chr == peaks_full$chr[1] |
#                    lodcolumn == peaks_full$lodcolumn[2] & chr == peaks_full$chr[2] |
#                    lodcolumn == peaks_full$lodcolumn[3] & chr == peaks_full$chr[3] |
#                    lodcolumn == peaks_full$lodcolumn[4] & chr == peaks_full$chr[4])
    
    # This gets the LOD over the marker with the highest LOD in the full dataset.
    df = data.frame(sim       = rep(sim, nrow(peaks_full)),
                    phenotype = peaks_full$lodcolumn, 
                    samples   = rep(ss, nrow(peaks_full)),
                    lod       = rep(0, nrow(peaks_full)),
                    pos       = peaks_full$pos,
                    prox      = rep(0, nrow(peaks_full)),
                    dist      = rep(0, nrow(peaks_full)))
    for(j in 1:nrow(df)) {
      df$lod[j]  = lod[peaks_full$marker[j], df$phenotype[j]]
      df$prox[j] = pk$ci_lo[pk$lodcolumn == df$phenotype[j]]
      df$dist[j] = pk$ci_hi[pk$lodcolumn == df$phenotype[j]]
    } # for(j)
    result = rbind(result, df)
    # This gets the maximum LOD on Chr 3.
#    result = rbind(result,
#                   data.frame(sim = sim, samples = ss, lod = pk$lod[1], pos = pk$pos[1]))
    sim = sim + 1
    
  } # for(i)

} # for(ss)
result = result[complete.cases(result),]
result$pos = result$pos * 1e-6
result$prox = result$prox * 1e-6
result$dist = result$dist * 1e-6
```

```{r}
library(tidyverse)
result %>%
#  mutate(phenotype = if_else(phenotype == 'cxcl1', 'Cxcl1', phenotype),
#         phenotype = if_else(phenotype == 'mtb_burden', 'M.tb. Burden', phenotype),
#         phenotype = if_else(phenotype == 's100a8', 'S100A8', phenotype),
#         phenotype = if_else(phenotype == 'tnf', 'TNF', phenotype)) %>%
  group_by(phenotype, samples) %>%
  mutate(mn = mean(lod, na.em = T),
         se = sd(lod, na.rm = T) / sqrt(n())) %>%
ggplot() +
  geom_point(aes(factor(samples), lod)) +
  geom_pointrange(aes(factor(samples), mn, ymin = mn - se, ymax = mn + se), color = 'red') +
  geom_hline(yintercept = 7.8, col = 'red') +
  facet_wrap(~phenotype, ncol = 2) +
  labs(x = 'Number of Samples', y = 'LOD')
```

Make high-res versions of each panel above and write them out.

```{r}
tmp = result %>%
         mutate(pheno2 = str_replace(phenotype, 'cxcl', 'Cxcl'),
                pheno2 = str_replace(pheno2, 'mtb_burden', 'M.tb. Burden'),
                pheno2 = str_replace(pheno2, 's100a8', 'S100A8'),
                pheno2 = str_replace(pheno2, 'tnf', 'TNF')) %>%
         group_by(phenotype, samples) %>%
         mutate(mn = mean(lod, na.em = T),
                se = sd(lod, na.rm = T) / sqrt(n()))

pretty_names = unique(tmp$pheno2)
names(pretty_names) = unique(tmp$phenotype)

for(p in unique(tmp$phenotype)) {
  png(file = file.path(base_dir, 'figures', paste0(p, '_power_sample_size.png')),
      width = 2000, height = 2000, res = 300)
  print(filter(tmp, phenotype == p) %>%
        ggplot() +
          geom_point(aes(factor(samples), lod), size = 2) +
          geom_pointrange(aes(factor(samples), mn, ymin = mn - se, ymax = mn + se), color = 'red', size = 1) +
          geom_hline(yintercept = 7.8, col = 'red') +
          labs(x = 'Number of Samples', y = 'LOD', title = pretty_names[p]) +
          theme(text = element_text(size = 24)))
  dev.off()
} # for(i)
```

Plot QTL width versus number of samples.

```{r}
result %>%
  mutate(phenotype = str_replace(phenotype, 'cxcl1', 'Cxcl1'),
         phenotype = str_replace(phenotype, 'mtb_burden', 'M.tb. Burden'),
         phenotype = str_replace(phenotype, 's100a8', 'S100A8'),
         phenotype = str_replace(phenotype, 'tnf', 'TNF'),
         interval  = dist - prox) %>%
  group_by(phenotype, samples) %>%
  mutate(mn = mean(interval, na.em = T),
         se = sd(interval, na.rm = T) / sqrt(n())) %>%
ggplot() +
  geom_point(aes(factor(samples), interval)) +
  geom_pointrange(aes(factor(samples), mn, ymin = mn - se, ymax = mn + se), color = 'red') +
  scale_y_log10() +
  facet_wrap(~phenotype, ncol = 2) +
  labs(x = 'Number of Samples', y = 'QTL Conf. Interval (Mb)')
```

```{r}
result %>%
  mutate(interval = dist - prox) %>%
  ggplot(aes(lod, interval)) +
     geom_point() +
     scale_y_log10() +
     facet_wrap(~phenotype)
```



Run permutations for different sample sizes.

```{r}
perms = NULL
nperm = 1000

for(ss in sample_size) {
  
    print(paste('sample', ss))
    
    samples = sample(rownames(pheno_rz), ss)
    pr = probs[samples,]
    ph = pheno_rz[samples, 's100a8', drop = FALSE]
    k  = K[[3]][samples, samples]
    cv = covar[samples,]
    
    p = scan1perm(genoprobs = pr, pheno = ph, kinship = k, addcovar = cv, n_perm = nperm, cores = 4)

    if(is.null(perms)) {
      perms = data.frame(p[,1])
    } else {
      perms = data.frame(perms, p[,1])
    } # else

} # for(ss)
colnames(perms) = sample_size
```

```{r}
apply(perms, 2, quantile, probs = 0.95)
```

Curious about coefs.

```{r coefs}
peaks = find_peaks(qtl_full, map, threshold = 7)
peaks$chr = as.character(peaks$chr)

for(i in 1:nrow(peaks)) {
  
  pheno_name = peaks$lodcolumn[i]
  chr = peaks$chr[i]
  
  print(paste(pheno_name, chr))
  
  blup = scan1blup(genoprobs = probs[,chr], pheno = pheno_rz[,pheno_name,drop = F], kinship = K[[chr]], addcovar = covar, cores = 4)
  
  plot_coefCC(blup, map, scan1_output = qtl_full[,pheno_name,drop = F], main = pheno_name)
  
} # for(i)
```

```{r}
snp_fxn  = create_variant_query_func(dbfile = '/media/dmgatti/hda/data/MUGA/cc_variants.sqlite')
gene_fxn = create_gene_query_func(dbfile    = '/media/dmgatti/hda/data/MUGA/mouse_genes_mgi.sqlite')
if(map[[1]][1] > 200) {
  map = lapply(map, '*', 1e-6)
}
for(i in 1:length(map)) {
  mkrs = intersect(names(map[[i]]), dimnames(probs[[i]])[[3]])
  map[[i]]   = map[[i]][mkrs]
  probs[[i]] = probs[[i]][,,mkrs]
} # for(i)
```

```{r}
ph = pheno_rz[,c('cxcl1', 'cxcl2', 'cxcl5', 's100a8')]
ph = ph[rowSums(is.na(ph)) < 4,]
ph_pca = pca(ph, nPcs = ncol(ph), scale = 'uv')

ph = matrix(scores(ph_pca)[,1], dimnames = list(rownames(ph), 'pc1'))
```

```{r pc1_chr15}
chr   = 15
start = 20
end   = 26
assoc = scan1snps(genoprobs = probs, map = map, pheno = ph, kinship = K, addcovar = covar, 
                  query_func = snp_fxn, chr = chr, start = start, end = end, cores = 4, keep_all_snps = TRUE)
```

```{r}
genes = gene_fxn(chr, start, end)
plot_snpasso(assoc$lod, assoc$snpinfo, genes = genes, top_panel_prop = 0.4, colors = 'black')
```

```{r cxcl5_chr8}
chr   = 8
start = 14
end   = 18
assoc = scan1snps(genoprobs = probs, map = map, pheno = ph, kinship = K, addcovar = covar, 
                  query_func = snp_fxn, chr = chr, start = start, end = end, cores = 4, keep_all_snps = TRUE)
```

```{r fig.width=10}
genes = gene_fxn(chr, start, end)
plot_snpasso(assoc$lod, assoc$snpinfo, genes = genes, top_panel_prop = 0.4, colors = 'black')
```

```{r s100a8_chr3}
chr   = 3
start = 90
end   = 93
assoc = scan1snps(genoprobs = probs, map = map, pheno = ph, kinship = K, addcovar = covar, 
                  query_func = snp_fxn, chr = chr, start = start, end = end, cores = 4, keep_all_snps = TRUE)
```

```{r fig.width=10}
genes = gene_fxn(chr, start, end)
plot_snpasso(assoc$lod, assoc$snpinfo, genes = genes, top_panel_prop = 0.4, colors = 'black')
```


