---
title: "Genoprobs QA/QC"
author: "Daniel Gatti"
date: "February 18, 2020"
output: 
  html_document:
      toc: true
      toc_depth: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(knitr)
library(qtl2)
library(readxl)
library(tidyverse)

base_dir = '/media/dmgatti/data0/Tufts/TB'
geno_dir = file.path(base_dir, "data/genotypes")
hap_dir  = file.path(base_dir, "haplo_reconstr")
gigamuga_url = 'https://raw.githubusercontent.com/kbroman/MUGAarrays/master/UWisc/gm_uwisc_v1.csv'
muga_dir = '/media/dmgatti/data0/MUGA'
pheno_file = file.path(base_dir, 'data', 'phenotypes', 'Beamer TB Phenotype data 2021-07-26.xlsx')
```

## Read in data

Load in the allele probs.

```{r read_probs}
aprobs = readRDS(file = file.path(hap_dir, "tufts_do_alleleprobs.rds"))
```

```{r read_pheno}
pheno = read_xlsx(pheno_file, sheet = 'Expt & phenotype data', na = 'na') %>% 
          select(mouse  = `Mouse #`,
                 coat   = `Coat color`,
                 strain = `Strain`,
                 gen    = `Generation`) %>% 
          mutate(mouse = as.character(mouse),
                 mouse = str_replace(mouse, '526.1', '526A'),
                 mouse = str_replace(mouse, '526.2', '526B'))
# Read in the covariates produced in the haplotype reconstruction pipeline.
covar = read_csv(file = file.path(hap_dir, "qtl2/covar.csv")) %>% 
          mutate(sample = as.character(sample),
                 sex    = 'F',
                 gen    = as.character(gen))
samples = readRDS(file.path(geno_dir, 'genotype_samples.rds'))
gmap = read.csv(file.path(hap_dir, 'qtl2', 'gmap.csv'))
gmap = qtl2convert::map_df_to_list(gmap, pos_column = 'pos')
```

## Genotyping Replicates

Find the samples that were run more than once.

```{r sample_dupl}
# Get the samples IDs and find duplicates.
mouse = strsplit(samples$iid, split = ';')
samples$mouse = sapply(mouse, '[', 2)
samples$mouse = sub('\\.[0-9]$', '', samples$mouse)
dupl = samples$mouse[which(duplicated(samples$mouse))]
samples$duplicated = samples$mouse %in% dupl

dupl = subset(samples, duplicated)
```

```{r compare_replicates,fig.width=10}
dupl_cor = matrix(0, nrow = nrow(dupl), ncol = nrow(dupl),
                  dimnames = list(dupl$iid, dupl$iid))

for(i in seq_along(aprobs)) {
  
  print(paste('chr', i))
  
  tmp = aprobs[[i]][dupl$iid,,]
  
  for(j in 1:dim(tmp)[3]) {
    dupl_cor = dupl_cor + cor(t(tmp[,,j]))
  } # for(j)
  
} # for(i)

dupl_cor = dupl_cor / rowSums(sapply(aprobs, dim))[3]

heatmap(dupl_cor)
```

The heatmap coloring above uses red for high and yellow for low. The large block in the upper right is a set of C57BL/6J mice. Their genoprobs have correlations near 1.0. 

There is one mouse with poor duplication: Sample 230. It was run in batches FPGM_14_15 and in Tufts_GIGMUGV01_20160622.

For most of the samples, we can use either run. But for sample 230, we need to select one of the replicates or discard the sample completely.

```{r}
sample_230_ids = filter(dupl, mouse == 230)
for(i in 1:nrow(sample_230_ids)) {
   qtl2::plot_genoprob(aprobs[sample_230_ids$iid[i],], map = gmap, main = sample_230_ids$iid[i])
}
```

I can't tell which one is correct right now.

## Founder allele probabilities for each sample.

Perform QC on the allele probs.

```{r founder_freq_per_sample}
fps = matrix(0.0, nrow(aprobs[[1]]), ncol(aprobs[[1]]))
for(p in aprobs) {
  fps = fps + apply(p, 1:2, sum)
} # for(p)
fps = fps / sum(sapply(aprobs, dim)[3,])

fps = cbind(fps, min = apply(fps, 1, min), max = apply(fps, 1, max))
write.csv(fps, file = file.path(base_dir, 'results', 'founder_freq_sample.csv'), quote = FALSE, row.names = TRUE)
```

Plot the founder allele frequencies for each sample.

```{r plot_sample_freq,fig.height=10,fig.width=10}
# Make new CC colors with a darker yellow.
cccolors2 = qtl2::CCcolors
cccolors2[1] = '#FFC800'
names(cccolors2) = LETTERS[1:8]

data.frame(fps) %>%
  rownames_to_column(var = "id") %>%
  select(-min, -max) %>%
  gather(founder, freq, -id) %>%
  ggplot() +
  geom_line(aes(id, freq, color = founder, group = founder)) +
  geom_abline(aes(intercept = 0.125, slope = 0)) +
  scale_color_manual(values = cccolors2) +
  facet_wrap(~founder, nrow = 8) +
  theme(axis.text.x = element_text(angle = 90, vjust = 1)) +
  labs(title = "Founder Allele Frequencies by Sample", y = "Founder Freq.")
```

We can clearly see the C57BL/6J samples (B probs ~ 1.0). Extract those and make sure that they're the ones that we expect.
 
Which mice are mostly BL6 in the founder probs, but aren't listed as so in the phenotypes?

```{r check_bl6}
bl6_in_fp = as.data.frame(fps) %>%
              rownames_to_column(var = 'id') %>% 
              mutate(mouse = samples$mouse) %>% 
              select(id, mouse, B) %>%
              filter(B > 0.9) %>%
              left_join(select(pheno, mouse, strain)) %>% 
              arrange(mouse)

kable(bl6_in_fp, digits = 3)
```

All of the mice are C57BL/6J above.

```{r check_bl6_all}
distinct(bl6_in_fp, strain)
```

Remove the mice that have BL6 contributions >= 0.9 from the founder probs.

```{r remove_bl6}
keep = setdiff(rownames(aprobs[[1]]), bl6_in_fp$id)

for(i in seq_along(aprobs)) {
  aprobs[[i]] = aprobs[[i]][keep,,]
} # for(i)

fps = fps[keep,]
```

Plot the founder allele frequency distribution for each sample.

```{r plot_sample_freq_density}
data.frame(fps) %>%
  rownames_to_column(var = "id") %>%
  select(-min, -max) %>%
  gather(founder, freq, -id) %>%
  ggplot() +
  geom_density(aes(freq, color = founder), size = 1) +
  geom_vline(aes(xintercept = 0.125), size = 1, linetype = 'dashed') +
  scale_color_manual(values = cccolors2) +
  theme(axis.text.x = element_text(angle = 90, vjust = 1)) +
  labs(title = "Density of Founder Allele Frequencies by Sample")
```

PWK (in red) is known to have a lower allele frequency than the other founders. The black dashed line in the center is the 1/8th (0.125) line.

## Founder Allele Probabilities along Chromosomes

Look at the founder allele frequencies along each chromosome.

```{r founder_freq_by_chr}
fpc = vector("list", length(aprobs))
names(fpc) = names(aprobs)
for(p in seq_along(aprobs)) {
  fpc[[p]] = apply(aprobs[[p]], 2:3, mean)
} # for(p)

tmp = t(fpc[[1]])
for(p in 2:length(fpc)) {
  tmp = rbind(tmp, t(fpc[[p]]))
} # for(p)
fpc = tmp
rm(tmp)

write.csv(fpc, file = file.path(base_dir, "results', 'founder_freq_marker.csv"), quote = FALSE, row.names = FALSE)
```

```{r plot_founder_freq_marker,fig.height=16,fig.width=8}
markers = readr::read_csv(file.path(muga_dir, 'gm_uwisc_v1.csv'), col_types = 'ccnnncclllnnc') %>%
            select(marker, chr, pos = bp_mm10)
data.frame(fpc) %>%
  rownames_to_column(var = "marker") %>%
  inner_join(y = select(markers, chr, marker, pos)) %>%
  mutate(pos = pos * 1e-6,
         chr = factor(chr, levels = c(1:19, 'X'))) %>%
  gather(founder, freq, -marker, -chr, -pos) %>%
  ggplot() +
  geom_line(aes(pos, freq, color = founder)) +
  geom_hline(aes(yintercept = 0.125), linetype = 2) +
  scale_color_manual(values = cccolors2) +
  facet_wrap(~chr, ncol = 1) +
  labs(title = "Founder Allele Frequency by Chromosome")
```

Some of the allele frequencies are quite high. On Chr X, 129S1 and NOD are high while WSB and PWK are low. PWK is somewhat expected. 

Plot the proportion of each founder by generation.

```{r founder_probs_by_gen}
samples = samples %>% 
            left_join(select(pheno, mouse, gen)) %>% 
            filter(iid %in% rownames(aprobs[[1]])) %>% 
            mutate(gen = str_replace(gen, '^G', ''),
                   gen = as.numeric(gen))

unique_gen = samples %>% 
               distinct(gen) %>% 
               filter(!is.na(gen)) %>% 
               arrange(gen) %>% 
               pull(gen)

fpr_gen = setNames(vector('list', length(unique_gen)), unique_gen)

for(g in unique_gen) {
  
  ss = samples$iid[samples$gen == g]
  
  fpc = setNames(vector("list", length(aprobs)), names(aprobs))
  for(p in seq_along(aprobs)) {
    fpc[[p]] = apply(aprobs[[p]][ss,,], 2:3, mean)
  } # for(p)

  tmp = t(fpc[[1]])
  for(p in 2:length(fpc)) {
    tmp = rbind(tmp, t(fpc[[p]]))
  } # for(p)
  fpr_gen[[as.character(g)]] = tmp
  rm(fpc, tmp)
  gc()
  
} # for(g)
```

```{r plot_g15,fig.height=16,fig.width=16}
df = data.frame(marker = rownames(fpr_gen[[1]]), fpr_gen[[1]], gen = names(fpr_gen)[1])
for(i in 2:length(fpr_gen)) {
  df = rbind(df, data.frame(marker = rownames(fpr_gen[[i]]), fpr_gen[[i]], gen = names(fpr_gen)[i]))
} # for(i)
df = left_join(df, markers, by = 'marker') %>% 
       mutate(pos = pos * 1e-6) %>% 
       pivot_longer(cols = A:H, names_to = 'founder', values_to = 'prop')

df %>% 
  ggplot() +
    geom_line(aes(pos, prop, color = gen)) + 
    scale_color_brewer(palette = 'YlGnBu') + 
    facet_grid(chr~founder) + 
    labs(title = 'Founder Proportions by Generation')
```


## Number of Recombinations per Sample.

```{r recomb_per_sample}
cross = qtl2::read_cross2(file = file.path(hap_dir, "qtl2/tufts.json"), quiet = FALSE)
cross = cross[keep,]
vit = qtl2:::viterbi2(cross = cross, quiet = FALSE, cores = 8)
saveRDS(vit, file = file.path(base_dir, "results/viterbi.rds"))

num_recomb = rep(0, nrow(vit[[1]]))
names(num_recomb) = rownames(vit[[1]])

for(v in vit) {
  vdiff = abs(diff(t(v)))
  num_recomb = num_recomb + colSums(vdiff > 0)
} # for(v)

num_recomb = data.frame(num_recomb) %>%
               rownames_to_column(var = "id")
```

```{r dist_num_recomb}
num_recomb %>%
  ggplot() +
  geom_density(aes(num_recomb)) +
  labs(title = "Distribution of Numer of Recombinations per Sample")
```

Plot the number of recombinations per sample by generation. We don't have generation data for the last set of mice yet.

```{r plot_recomb_per_sample}
num_recomb = num_recomb %>%
               mutate(mouse = str_replace(id, '^(neo|jax|unc)_', '')) %>%
               left_join(covar, by = c('mouse' = 'sample'))

num_recomb %>%
  mutate(gen = as.numeric(gen)) %>% 
  ggplot() +
  geom_boxplot(aes(gen, num_recomb, group = gen)) +
  geom_jitter(aes(gen, num_recomb), alpha = 0.3) +
  geom_smooth(aes(gen, num_recomb), method = 'lm') +
  labs(title = "Number of Recombinations per Sample") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 0.5))
```

This looks reasonable. It's interesting to see how many recombinations the later generations have. I wonder if the Gigamuga can detect these accurately? The line still appears to be going up linearly.


## Coat Color Checks

The Albino locus is on Chr 7 at 87.42 - 87.49 Mb (Tyr) and the Agouti locus is on Chr 2 at 154.79 - 155.05 Mb (a). Mice carrying the A and D alleles at the albino locus will be albino. Mice carrying the A or B alleles on Chr 2 that aren't albino will be black. We can compare the coat color in the phenotype data with the founder allele at the albino locus to determine if there has been a sample mixup.

Remove the C57BL/6J mice and change the coat colors to wh, br & bl.

```{r}
pheno = pheno %>% 
          filter(coat != 'C57BL/6J' & coat != 'unknown')
pheno = pheno %>%
          mutate(mouse = as.character(mouse),
                 coat = if_else(coat == 'w', 'wh', coat),
                 coat = str_sub(coat, 1, 2),
                 coat = if_else(startsWith(coat, 'na'), NA_character_, coat),
                 coat = str_trim(coat))
count(pheno, coat)
```

Get the founder allele calls on Chr 7 near the Tyr locus.

```{r get_tyr_gt}
# Markers near Tyr gene.
tyr_mkr = subset(markers, chr == '7' & pos >= 87400000 & pos <= 87500000)
tyr_mkr = tyr_mkr[tyr_mkr$marker %in% dimnames(aprobs[[7]])[[3]],]
# Markers near a gene.
a_mkr   = subset(markers, chr == '2' & pos >= 154790000 & pos <= 155100000)
a_mkr   = a_mkr[a_mkr$marker %in% dimnames(aprobs[[2]])[[3]],]
# Get the probs for the agouti locus.
pr_2 = aprobs[[2]][,,a_mkr$marker]
# Get the probs for the albino locus.
pr_7 = aprobs[[7]][,,tyr_mkr$marker]

samples = sub('\\.[0-9]$', '', rownames(aprobs[[1]]))
# Add 526.1 and 526.2 back in.
samples[samples == '526'] = c('526.1', '526.2')

# For each sample, list the sample ID, listed coat color, prob of A/D and A/B genotypes.
samples = data.frame(geno_id = rownames(pr_7), sample = samples, coat = NA, prob_AD = NA, prob_AB = NA, remove = F)

for(i in 1:nrow(samples)) {
  
  # Get the phenotype row for this sample.
  pheno_wh = which(pheno$mouse == samples$sample[i])
  
  # If the sample was found, match up coat color with genotype probs.
  if(length(pheno_wh) > 0) {
    # Assign coat color.
    samples$coat[i] = pheno$coat[pheno_wh]
    # Assign A/D prob.
    samples$prob_AD[i] = sum(rowMeans(pr_7[i,,])[c('A', 'D')])
    # Assign A/B prob.
    samples$prob_AB[i] = sum(rowMeans(pr_2[i,,])[c('A', 'B')])
  } # if(length(wh) > 0)
  
} # for(i)
```

Which samples were listed as having a white coat in the phenotype file but don't have an AA, AD or DD genotype?

```{r which_coat_non_geno}
coat_mismatch = filter(samples, coat == 'wh' & prob_AD < 0.9) %>% 
                arrange(geno_id)
coat_mismatch
```

Seven samples: 375, 699, 701, 943, 952, 960, 965.

Which samples have AA, AD or DD genotypes but aren't listed as 'white' in the phenotype file?

```{r white_geno_not_pheno}
coat_mismatch = bind_rows(coat_mismatch, filter(samples, coat != 'wh' & prob_AD > 0.9))
coat_mismatch
```

Eight samples: 230, 626, 640, 669, 673, 951.

Which samples were listed as having a black coat in the phenotype file but don't have an AA, AB or BB genotype?

```{r black_coat_non_geno}
coat_mismatch = bind_rows(coat_mismatch, filter(samples, coat == 'bl' & prob_AB < 0.9 & prob_AD < 0.9))
coat_mismatch
```

Four samples: 333, 692, 955, 957.

Which samples have AA, AB or BB genotypes and are not albino but aren't listed as 'black' in the phenotype file?

```{r black_geno_not_pheno}
coat_mismatch = bind_rows(coat_mismatch, filter(samples, coat != 'bl' & prob_AB > 0.9 & prob_AD < 0.9))
coat_mismatch
```

Nine samples: 332, 701, 867, 943, 952, 953, 956, 959, 960.

```{r write_coat_color}
coat_mismatch = coat_mismatch %>% 
                  mutate(across(.cols = prob_AD:prob_AB, .fns = ~ round(.x, digits = 2))) %>% 
                  rename(num_albino_alleles = prob_AD,
                         num_black_alleles  = prob_AB) %>% 
                  filter(startsWith(geno_id, 'neo')) %>% 
                  select(sample:num_black_alleles)

# Add the sample batch.
tufts_dirs = list.dirs(geno_dir, recursive = FALSE)
tufts_dirs = tufts_dirs[grep('/Tufts_|FPGM', tufts_dirs)]

geno_samples = NULL
for(d in tufts_dirs) {
  sample_file = dir(d, pattern = '^Sample', full.names = T)
  s = read_delim(sample_file[1], delim = '\t') %>% 
        mutate(Name = as.character(Name),
               ID = as.character(ID),
               batch  = sub(paste0(geno_dir, '/'), '', d))
  geno_samples = bind_rows(geno_samples, s)
} # for(d)

geno_samples = geno_samples %>% 
                 select(sample = ID, batch)

coat_mismatch = left_join(coat_mismatch, geno_samples, by = 'sample') %>% 
                mutate(num_albino_alleles = 2 * num_albino_alleles,
                       num_black_alleles  = 2 * num_black_alleles) %>% 
                distinct() %>% 
                arrange(sample)

write_csv(coat_mismatch, file = file.path(hap_dir, 'tufts_do_coat_color_check_20210805.csv'))
```


Maybe the haplotype reconstruction is not perfect at the later generations and the HMM isn't transitioning to the correct state. Many of the albino mice have 0.5 A or D at the Tyr locus. 

For 230, the sample in batch 3 differs from the other batches and does not have the correct coat color. I will remove the 'neo_230' sample.
Samples 332 & 333 and 916 & 918 are most likely swapped. I will swap them in the data. 
Samples 375, 643, 943,  do not have the correct coat color. I will remove them. 

```{r remove_coat_mismatch}
#remove = c('neo_230', 'neo_375', 'jax_375', 'neo_514', 'neo_626', 'neo_640', 'neo_673', 'neo_701')
remove = c('jax_375', paste0('neo_', coat_mismatch$sample))
remove = remove[!remove %in% c('neo_332', 'neo_333')]

for(i in seq_along(aprobs)) {
  
  # Remove bad coat color samples.
  aprobs[[i]] = aprobs[[i]][!rownames(aprobs[[i]]) %in% remove,,]
  
  # Swap 332 & 333.
  tmp                      = aprobs[[i]]['neo_332',,]
  aprobs[[i]]['neo_332',,] = aprobs[[i]]['neo_333',,]
  aprobs[[i]]['neo_333',,] = tmp

} # for(i)
```

## Remove Duplicates

Keep only unique samples and write out the cleaned founder allele probs.

```{r remove_dupl}
samples = data.frame(geno_id = rownames(aprobs[[1]]),
                     sample  = NA,
                     dupl    = F)
samples$sample = sub('^(neo|jax|unc)_', '', samples$geno_id)
samples$sample = sub('\\.[0-9]$',       '', samples$sample)
samples$dupl = duplicated(samples$sample)
keep = which(!samples$dupl)

for(i in seq_along(aprobs)) {

  aprobs[[i]] = aprobs[[i]][keep,,]
  rownames(aprobs[[i]]) = samples$sample[keep]

} # for(i)
```

Dimensions of allele probs. Each element represents one chromosome with dimensions of samples, by founders by markers.

```{r}
sapply(aprobs, dim)
```

We have 783 samples.

Write out the cleaned founder allele probs.

```{r write_cleaned_probs}
# Sort the sample IDs.
new_order = order(rownames(aprobs[[1]]))
for(i in seq_along(aprobs)) {

  aprobs[[i]] = aprobs[[i]][new_order,,]

} # for(i)

saveRDS(aprobs, file = file.path(hap_dir, 'tufts_do_alleleprobs_cleaned.rds'))
```

