---
title: "Tufts Haplotype Reconstruction"
author: "Daniel Gatti"
date: "December 30, 2019"
output: html_document
---

```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = FALSE)
library(knitr)
library(readxl)
library(argyle)
library(tidyverse)
library(qtl2)
library(qtl2convert)

base_dir = "/media/dmgatti/data0/Tufts/TB/"
geno_dir = paste0(base_dir, "data/genotypes/")
hap_dir  = paste0(base_dir, "haplo_reconstr/")
gigamuga_url = 'https://raw.githubusercontent.com/kbroman/MUGAarrays/master/UWisc/gm_uwisc_v1.csv'
muga_dir = '/media/dmgatti/data0/MUGA/'
```

Karl Broman at Univ. of Wisconsin [https://kbroman.org/MUGAarrays/new_annotations.html](remapped the GigaMUGA markers) and has produced a new  [https://raw.githubusercontent.com/kbroman/MUGAarrays/master/UWisc/mini_uwisc_v1.csv](marker annotation file).

```{r read_markers}
# Read in the U. Wisc. remapped markers.
markers = readr::read_csv(str_c(muga_dir, 'gm_uwisc_v1.csv'), col_types = 'ccnnncclllnnc') %>%
          filter(!is.na(chr)) %>%
          as.data.frame()
colnames(markers)[3:4] = c('pos', 'cM')
rownames(markers) = markers$marker

map  = qtl2convert::map_df_to_list(map = markers, pos_column = 'pos')
gmap = qtl2convert::map_df_to_list(map = markers, pos_column = 'cM')

markers = markers[,c(2,1,4,3)]

# Also load UNC markers.
load(paste0(muga_dir, "snps.gigamuga.Rdata"))
```

Check the marker name overlap. All of the markers in 'marker' are in 'snps' from UNC.

```{r check_markers}
all(rownames(markers) %in% rownames(snps))
```

Load in the Geneseek data. There are files that start with 'FPGM' that are from UNC and contain data from other investigators. Gillian's samples have 'GB' and/or 'Tufts' in their names. The directories that begin with 'Tufts' are all Gillian's samples.

```{r gather_geno}
data_dirs = dir(geno_dir, pattern = '^(Tufts_|FPGM)', full.names = TRUE)

# Variables for genotypes and samples.
geno    = NULL
samples = NULL

# Read in all of the data and merge into one large object.
for(d in data_dirs) {
  
  print(d)
  
  ds = unlist(strsplit(d, split = "/"))
  ds = ds[length(ds)]
  
  # Read in data from Neogen file.
  tmp = read.beadstudio(prefix = "ds__", snps = snps, in.path = d)
  
  # Subset to keep Gillian's samples in UNC run files.
  if(startsWith(ds, 'FPGM')) {
    tmp = subset(tmp, grepl('Tufts', samples(tmp)$iid), by = 'samples')
    tmp = replace.names(gty = tmp, nn = gsub('^Tufts_GB(2_|_|07\\.2J\\:DO)?', '', samples(tmp)$iid))
  } # if(startsWith(d, 'FPGM'))
  
  # Tufts_MURGIGV01_20151118: Ignore allele calls for 173-184. Beamer e-mail.
  if(ds == 'Tufts_MURGIGV01_20151118') {
    
    tmp = subset(tmp, !samples(tmp)$iid %in% as.character(173:184), by = 'samples')
    tmp = subset(tmp, !samples(tmp)$iid %in% c('5925011', '5925016', '5925021'), by = 'samples')
    
  } # if(ds == 'Tufts_MURGIGV01_20151118')
  
  # 20160818_Fixed  12 (173-178 J:DO; 179-184 C57BL/6J). Beamer e-mail.
  if(ds == '20160818_Fixed') {
    
    tmp = subset(tmp, !samples(tmp)$iid %in% as.character(179:184), by = 'samples')
    
  } # if(ds == '20160818_Fixed')

  # Save sample IDs
  samples = rbind(samples, cbind(ds, samples(tmp)))
  
  # Add batch ID to sample IDs to make unique sample names.
  tmp = replace.names(gty = tmp, nn = paste(ds, samples(tmp)$iid, sep = ";"))

  if(is.null(geno)) {
    geno = tmp
  } else {
    geno = cbind.genotypes(geno, tmp) 
  } # else

} # for(d)
samples$iid = colnames(geno)
```

Print out the number of samples in each batch.

```{r num_samples_per_batch}
data.frame(table(samples$ds))
```

Check whether we have ~143000 rows.

```{r}
dim(geno)
```

Filter genotypes to remove those with a on no-call rate > 10%.

```{r}
geno = as.data.frame(geno)
rownames(geno) = geno$marker
geno = as.matrix(geno[,-(1:6)])
call_rate = colMeans(geno == 'N', na.rm = TRUE)
outliers = call_rate[call_rate > 0.1]
outliers = data.frame(sample = names(outliers), call_rate = outliers)
kable(outliers)
write.csv(outliers, file = file.path(base_dir, 'results', 'genotype_outliers.csv'))
geno = geno[,call_rate <= 0.1]
samples = samples[call_rate <= 0.1,]
stopifnot(nrow(samples) == ncol(geno))
```

Make a histogram of the genotypes call rates.

```{r call_rate_histo}
hist(1.0 - call_rate, breaks = 100, xlim = c(0.9, 1.0))
```

Select UNC tier 1 & 2 markers and ones in Karl's marker set.

```{r select_markers}
snps = snps[snps$tier <= 2,]
markers = markers[markers$marker %in% snps$marker,]
marker_set = intersect(rownames(geno), markers$marker)
markers = markers[markers$marker %in% marker_set,]
geno = geno[markers$marker,]
```

Fix some of the sample IDs.

```{r fix_sample_ids}
colnames(geno) = sub(';3\\.1$',  ';23', colnames(geno))
colnames(geno) = sub(';32\\.1$', ';33', colnames(geno))
colnames(geno) = sub(';78\\.1$', ';79', colnames(geno))
samples$iid = sub(';3\\.1$',  ';23', samples$iid)
samples$iid = sub(';32\\.1$', ';33', samples$iid)
samples$iid = sub(';78\\.1$', ';79', samples$iid)
samples$fid = sub('^3\\.1$',  ';23', samples$fid)
samples$fid = sub('^32\\.1$', ';33', samples$fid)
samples$fid = sub('^78\\.1$', ';79', samples$fid)
```

Save the genotypes.

```{r save_geno}
saveRDS(samples, file = file.path(geno_dir, 'genotype_samples.rds'))
saveRDS(geno,    file = file.path(geno_dir, 'genotypes.rds'))
gc()
```

Look at duplicate samples.

```{r duplicate_samples}
# Add a column with the mouse ID.
tmp = sub('\\.[0-9]$', '', samples$iid)
tmp = strsplit(tmp, ';')
samples$mouse = sapply(tmp, '[', 2)

# For each sample that was run more than once, compare the genotypes.
# If they are largely identical, get consensus values.
# If they are not, flag them.
samples$dupl = duplicated(samples$mouse)
samples$dupl = samples$mouse %in% samples$mouse[samples$dupl]
sample_ct = sort(table(samples$mouse[samples$dupl]))

x = rep(0, length(sample_ct))

for(i in seq_along(sample_ct)) {
  
  mouse = names(sample_ct[i])
  ids = samples$iid[samples$mouse == mouse]
  
  curr_geno = geno[,ids]
  
  x[i] = mean(curr_geno[,1] == curr_geno[,2])
  if(x[i] < 0.9) { 
    print(apply(curr_geno, 2, table))
  } # if(x[i] < 0.9)
  
} # for(i)

```

Only one sample seems to have poor duplication. Mouse 230 is agouti, so we may be able to tell which sample is the correct once from the haplotype probs.


Save the sample sex.

```{r save_sex}
# All of the samples are female.
sex = data.frame(id = colnames(geno),
                 sex = 'F')

write.csv(sex, file = paste0(hap_dir, "qtl2/sex.csv"), quote = FALSE, row.names = FALSE)
```

Gather the founder and sample genotypes and convert them to qtl2 format.

```{r get_geno_calls}
load(file.path(muga_dir, 'jax_ftp', "GM_geno.Rdata"))
load(file.path(muga_dir, 'jax_ftp', "GM_code.Rdata"))
load(file.path(muga_dir, 'jax_ftp', "GM_sex.Rdata"))

# Fix B6 markers.
b6 = grep('^B6', rownames(GM_geno))
rownames(GM_geno)[b6] = gsub('\\.', '-', rownames(GM_geno)[b6])

# Reorder to match the current SNPs.
GM_geno = GM_geno[rownames(markers),]
stopifnot(rownames(GM_geno) == rownames(markers))

stopifnot(rownames(GM_geno) == rownames(markers))
stopifnot(rownames(geno) == rownames(markers))
```


```{r get_consensus_calls}
# Keep only the founders.
founder_codes = paste0(LETTERS[1:8], LETTERS[1:8])
keep = which(GM_code %in% founder_codes)
GM_geno = GM_geno[,keep]
GM_code = GM_code[keep]
GM_sex  = GM_sex[keep]
GM_geno[is.na(GM_geno)] = 'N'

ug = qtl2convert::find_unique_geno(GM_geno)
sum(is.na(ug[,1]))

fgeno = matrix("", nrow = nrow(GM_geno), ncol = length(founder_codes), dimnames = list(rownames(GM_geno), founder_codes))

for(i in founder_codes) {
  
  fgeno[,i] = qtl2convert::find_consensus_geno(GM_geno[,GM_code == i])
  
} # for(i)

colnames(fgeno) = substring(colnames(fgeno), 1, 1)
```

```{r encode_genotypes}
# This has the undesired side-effect of sorting the rownames.
comb_geno = merge(fgeno, geno, by = "row.names", sort = FALSE)
rownames(comb_geno) = comb_geno[,1]
comb_geno = as.matrix(comb_geno[,-1])
comb_geno = comb_geno[rownames(markers),]

new_geno = comb_geno

ug = apply(comb_geno, 1, table)

for(i in 1:nrow(comb_geno)) {
  
  # N
  wh = which(comb_geno[i,] == "N")
  new_geno[i, wh] = "-"
  
  ug[[i]] = ug[[i]][!names(ug[[i]]) %in% c("H", "N")]
  stopifnot(length(ug[[i]]) < 3)
  
  # Allele A
  if(length(ug[[i]]) > 0) {
    wh = which(comb_geno[i,] == names(ug[[i]])[1])
    new_geno[i, wh] = "A"
  }
  
  # Allele B
  if(length(ug[[i]]) > 1) {
    wh = which(comb_geno[i,] == names(ug[[i]])[2])
    new_geno[i, wh] = "B"
  }
  
} # for(i)

# Write founders.
tmp = data.frame(marker = rownames(new_geno), new_geno[,1:8])
write.csv(tmp, file = paste0(hap_dir, "qtl2/founder_geno.csv"), quote = FALSE, row.names = FALSE)

# Write founders.
tmp = data.frame(marker = rownames(new_geno), new_geno[,-(1:8)])
colnames(tmp) = sub('^X', '', colnames(tmp))
colnames(tmp) = sub('\\.', ';', colnames(tmp))
write.csv(tmp, file = paste0(hap_dir, "qtl2/sample_geno.csv"), quote = FALSE, row.names = FALSE)
rm(tmp)
```

```{r gmap_pmap}
# Write out genetic and physical maps.
tmp = markers[,c("marker", "chr", "cM")]
colnames(tmp)[3] = "pos"
write.csv(tmp, file = paste0(hap_dir, "qtl2/gmap.csv"), quote = FALSE, row.names = FALSE)
rm(tmp)

tmp = markers[,c("marker", "chr", "pos")]
tmp$pos = tmp$pos * 1e-6
write.csv(tmp, file = paste0(hap_dir, "qtl2/pmap.csv"), quote = FALSE, row.names = FALSE)
rm(tmp)
```

```{r make_pheno_covar}
pheno = readxl::read_xlsx(path = file.path(base_dir, 'data', 'phenotypes', 'Beamer TB Phenotype data 2021-07-26.xlsx'),
                          sheet = 'Expt & phenotype data', na = 'na')
pheno = pheno[,1:5]
colnames(pheno) = c('mouse', 'coat', 'strain', 'gen', 'expt')
pheno$mouse = as.character(pheno$mouse)
pheno$mouse[pheno$mouse == '526.1'] = '526A'
pheno$mouse[pheno$mouse == '526.2'] = '526B'
pheno = as.data.frame(pheno)
rownames(pheno) = pheno$mouse

# Map the genotype samples to the phenotype samples.
sample_int = intersect(pheno$mouse, samples$mouse)

pheno = pheno[sample_int,]
pheno = pheno[samples$mouse,]
rownames(pheno) = colnames(geno)

covar = data.frame(id     = rownames(pheno),
                   sample = rownames(pheno),
                   gen    = pheno$gen,
                   sex    = 'F')             # All samples are female.
covar$gen = sub('^G', '', covar$gen)
covar$gen[is.na(covar$gen)] = '1'
covar$gen = as.numeric(covar$gen)
rownames(covar) = covar$id

# Create fake phenotypes just for haplotype reconstruction.
pheno = data.frame(id = covar$id, junk = runif(nrow(covar)))
rownames(pheno) = pheno$id

write.csv(pheno, file = file.path(hap_dir, "qtl2/pheno.csv"), quote = TRUE, row.names = FALSE)
write.csv(covar, file = file.path(hap_dir, "qtl2/covar.csv"), quote = TRUE, row.names = FALSE)
```

```{r cleanup}
rm(list = ls())
gc()
```

################################################
# START HERE ONCE ABOVE HAS BEEN RUN.
################################################

```{r load_libs}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = FALSE)
library(knitr)
library(qtl2)
library(qtl2convert)

base_dir = "/media/dmgatti/data0/Tufts/TB/"
geno_dir = paste0(base_dir, "data/genotypes/")
hap_dir  = paste0(base_dir, "haplo_reconstr/")
gigamuga_url = 'https://raw.githubusercontent.com/kbroman/MUGAarrays/master/UWisc/gm_uwisc_v1.csv'
muga_dir = '/media/dmgatti/data0/MUGA/jax_ftp/'
```

Read in the DO cross.

```{r read_cross}
cross = qtl2::read_cross2(file = paste0(hap_dir, "qtl2/tufts.json"), quiet = FALSE)
```

```{r estimate_haplotype_probs}
# Run haplotype reconstruction.
probs = qtl2:::calc_genoprob2(cross = cross, quiet = FALSE, cores = 8)
saveRDS(probs, file = paste0(hap_dir, "tufts_do_genoprobs.rds"))

gc()

# Convert to allele probs.
aprobs = genoprob_to_alleleprob(probs = probs, quiet = FALSE, cores = 4)
saveRDS(aprobs, file = paste0(hap_dir, "tufts_do_alleleprobs.rds"))

rm(probs)
gc()
```
