---
title: "Tufts DO TB Phenotype QA/QC"
author: "Daniel Gatti"
date: "February 20, 2020"
output: html_document
---

```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(readxl)
library(corrplot)
library(GGally)
library(ggrepel)
library(tidyverse)
base_dir = '/media/dmgatti/data0/Tufts/TB'
pheno_file = 'Beamer TB Phenotype data 2021-09-02.xlsx'
```

# Initial Data Cleanup

I am focusing on phenotype cleanup for the DO mice to pass into the QTL mapping scripts.

Read in phenotype data.

```{r read_pheno}
pheno = readxl::read_xlsx(file.path(base_dir, 'data', 'phenotypes', pheno_file), sheet = 'Expt & phenotype data', na = 'na')
```

What are the data dimensions?

```{r pheno_dim}
dim(pheno)
```

What are the column names?

```{r pheno_colnames}
colnames(pheno)
```

Make the column names shorter.

```{r fix_colnames}
pheno_orig_colnames = colnames(pheno)

pheno = pheno %>% 
          select(mouse       = `Mouse #`,
                 coat        = `Coat color`,
                 strain      = `Strain`,
                 gen         = `Generation`,
                 expt        = `Experiment`,
                 mach_type   = `Aerosol machine`,
                 gigamuga    = `GigaMUGA`,
                 ge_array    = `Lung Gene Expression (Microarray)`,
                 slide_scan  = `Slide scanned`,
                 cage        = `Cage #`,
                 aero_run    = `Aerosol run no`,
                 mtb_dose    = `Mtb initial dose (average)`,
                 susc_class  = `Class`,
                 il10_assay    = `IL-10 Assay Source`,
                 il12_assay    = `IL-12 Assay Source`,
                 euth_day     = `Day euth Post-Infection`,
                 euth         = `Euth due to predetermined timepoint (y/n)`,
                 map_survival = `QTL map survival (y/n)`,
                 map_lung     = `QTL map lung traits (y/n)`,
                 pct_wt_loss = `% wt loss cf peak`,
                 mtb_burden  = `Lung Mtb burden`,
                 lung_cxcl5  = `Lung CXCL5 pg/ml`,
                 lung_cxcl2  = `Lung CXCL2 pg/ml`,
                 lung_cxcl1  = `Lung CXCL1 pg/ml`,
                 lung_tnf    = `Lung TNF pg/ml`,
                 lung_mmp8   = `Lung MMP8 pg/ml`,
                 lung_s100a8 = `Lung S100A8 pg/mL`,
                 lung_s100a9 = `Lung S100A9 pg/mL`,
                 lung_resistin = `Lung resistin pg/ml`,
                 lung_il1a     = `Lung IL-1a pg/mL`,
                 lung_il1b     = `Lung IL-1B pg/mL`,
                 lung_il6      = `Lung IL-6 pg/mL`,
                 lung_ifng     = `Lung IFNg pg/ml`,
                 lung_il12     = `Lung IL-12 pg/ml`,
                 lung_il17     = `Lung IL-17 pg/ml`,
                 lung_il10     = `Lung IL-10 pg/ml`,
                 lung_il4      = `Lung IL-4 pg/mL`,
                 lung_vegf     = `Lung VEGF pg/ml`,
                 lung_ccl3     = `Lung CCL3 pg/ml`,
                 lung_ccl4     = `Lung CCL4 pg/ml`,
                 lung_ccl5     = `Lung CCL5 pg/ml`,
                 lung_anti_esat6_cfp10 = `Lung anti-ESAT6:CFP10 pg/mL`,
                 lung_anti_mtb_cw_igg  = `Lung anti-M.tb CW IgG \"ng/mL\"`,
                 lung_anti_mtb_cfp_igg = `Lung anti-M.tb CFP IgG \"ng/mL\"`,
                 lung_anti_crypto_igg  = `Lung anti-crypto IgG \"ng/mL\"`,
                 blood_esat6_ifng      = `Blood ESAT-6 specific IFNg pg/ml`,
                 lung_prob_SS          = `Lung granuloma Probability of SS`,
                 lung_necr_ratio       = `Lung granuloma Ratio of pyknotic necrosis to non-necrosis`,
                 lung_stain            = `Lung stain`,
                 blood_esat6_tnf       = `Blood ESAT-6 specific TNF pg/ml`,
                 blood_esat6_il12p70   = `Blood ESAT-6 specific IL-12p40/p70 pg/ml`,
                 blood_esat6_il10      = `Blood ESAT-6 specific IL-10 pg/ml`,
                 blood_esat6_il2       = `Blood ESAT-6 specific IL-2 pg/ml`,
                 blood_pct_neut        = `% blood neutrophils`,
                 serum_cxcl5           = `Serum CXCL5 pg/ml`,
                 serum_cxcl2           = `Serum CXCL2 pg/ml`,
                 serum_cxcl1           = `Serum CXCL1 pg/ml`,
                 serum_tnf             = `Serum TNF pg/ml`,
                 seum_mmp8             = `Serum MMP8 pg/ml`,
                 serum_s100a8          = `Serum S100A8 pg/ml`,
                 serum_resistin        = `Serum resistin pg/ml`,
                 serum_il1b            = `Serum IL-1B pg/ml`,
                 serum_il6             = `Serum IL-6 pg/mL`,
                 serum_ifng            = `Serum IFNg pg/mL`,
                 serum_il12p70         = `Serum IL-12p70 pg/ml`,
                 serum_il17            = `Serum IL-17 pg/ml`,
                 serum_il10            = `Serum IL-10 pg/mL`,
                 serum_il4             = `Serum IL-4 pg/mL`,
                 serum_vegf            = `Serum VEGF pg/ml`,
                 serum_ccl3            = `Serum CCL3 pg/ml`,
                 serum_ccl4            = `Serum CCL4 pg/ml`,
                 serum_ccl5            = `Serum CCL5 pg/ml`,
                 serum_anti_mtb_cw_igg = `Serum anti-M.tb CW IgG \"ng/mL\"`,
                 serum_anti_mtb_cfp_igg = `Serum anti-M.tb CFP IgG \"ng/mL\"`,
                 serum_anti_crypto_igg = `Serum anti-crypto IgG \"ng/mL\"`)
```

Remove columns that are all NA. There are `r ncol(pheno)` columns before.

```{r remove_na}
orig_pheno = pheno
orig_pheno$mouse = as.character(orig_pheno$mouse)

pheno = bind_cols(select(pheno, mouse:euth),
                  pheno %>% 
                    select(pct_wt_loss:lung_necr_ratio, blood_esat6_tnf:serum_anti_crypto_igg) %>% 
                    mutate(across(.cols = where(is.character), .fns = parse_number)))

pheno = pheno[,colMeans(is.na(pheno)) < 1.0]
```

There are `r ncol(pheno)` columns after.

Clean up data columns. 

```{r data_clean1}
pheno = pheno %>%
          mutate(mouse = as.character(mouse),
                 gen   = if_else(gen == 'na' | gen == 'NA', NA_character_, gen),
                 mach_type = if_else(str_detect(mach_type, '^na'), NA_character_, mach_type),
                 gigamuga  = if_else(str_detect(gigamuga, '^y'), 'y', gigamuga),
                 gigamuga  = if_else(str_detect(gigamuga, 'na'), NA_character_, gigamuga),
                 euth      = if_else(str_detect(euth, '^n'), 'n', euth),
                 euth      = if_else(str_detect(euth, '^y'), 'y', euth)) %>% 
         filter(str_detect(mouse, '526', negate = TRUE))
```


Make sure that there are no duplicated mouse IDs.

```{r}
count(pheno, mouse) %>% distinct(n)
```


At this point, the columns have been renamed and the columns contain characters or numbers, as appropriate.

# Missing Data

Keep only the DO mice.

```{r get_numeric_pheno}
pheno = pheno %>%
           filter(strain == 'J:DO')
pheno_num = select(pheno, mouse, where(is.numeric)) %>% 
            select(-(cage:mtb_dose))
```

How many missing (i.e. NA) data points are there in each column?

```{r missing_data}
missing = pheno_num %>% 
            pivot_longer(cols = pct_wt_loss:serum_anti_crypto_igg, names_to = 'phenotype') %>% 
            mutate(is_na = is.na(value)) %>% 
            group_by(phenotype) %>% 
            summarize(pct_na = mean(is_na) * 100.0,
                      not_na = sum(!is_na))
kable(missing, digits = 3)
```

Keep columns with > 300 samples.

```{r complete_data}
columns2keep = c(colnames(pheno)[1:17], filter(missing, not_na >= 300) %>% pull(phenotype))
pheno = pheno %>% 
          select(which(colnames(pheno) %in% columns2keep))
```

The sample size is very small for some phenotypes. 60 to 110 mice will be too small to map anything reliably. The phenotypes with over 300 samples will be more promising.

```{r missing_pattern,fig.height=8,fig.width=12}
pheno_num %>%
  bind_cols(pull(pheno, mouse)) %>%
  pivot_longer(cols = pct_wt_loss:serum_anti_crypto_igg, names_to = 'phenotype', values_to = 'missing') %>%
  mutate(missing = is.na(missing),
         missing = factor(missing, levels = c(TRUE, FALSE))) %>%
  ggplot() +
    geom_tile(aes(x = mouse, y = phenotype, fill = missing)) +
    scale_fill_grey() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```

The missing data is related to batches of samples. There are blocks of samples where we have a broad set of phenotypes. We'll have to think about which phenotypes have enough samples to use for correlation analysis.

# Univariate Distributions

Here, I look at the distribution of each phenotype to determine if it will require transformation before analysis and to look for outliers. 

Look at the range of data in each column.

```{r pheno_range}
pheno_num %>% 
  select(euth_day, pct_wt_loss:serum_anti_crypto_igg) %>% 
  rename_all(str_replace_all, pattern = '_', replacement = '.') %>%
  summarise_all(list(min, max, mean, median), na.rm = TRUE) %>%
  pivot_longer(cols = everything(), names_to = 'phenotype', values_to = 'value') %>%
  separate(phenotype, into = c('phenotype', 'fxn'), sep = '_') %>%
  mutate(fxn = str_replace(fxn, 'fn1', 'min'),
         fxn = str_replace(fxn, 'fn2', 'max'),
         fxn = str_replace(fxn, 'fn3', 'mean'),
         fxn = str_replace(fxn, 'fn4', 'median')) %>%
  pivot_wider(names_from = fxn, values_from = value) %>%
  select(phenotype, min, max, mean, median) %>%
  kable(digits = 1)
```

Some phenotypes have values of 0, which means that we will have to add one before taking the log. Pct_wt_loss has negative values, which I assume means that the mouse gained weight.

Initial boxplot of log+1 transformed values.

```{r pheno_boxplot, fig.width=12}
pheno_num %>% 
  select(euth_day, pct_wt_loss:serum_anti_crypto_igg) %>% 
  pivot_longer(cols = everything(), names_to = 'phenotype', values_to = 'value') %>%
  mutate(value = log1p(value)) %>%
  ggplot(aes(phenotype, value)) +
    geom_boxplot() +
    geom_jitter(alpha = 0.3) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```

Most of these look good. A few have zero values when all of the other mice have positive values. Is this due to a detection limit issue?

Standardize the phenotypes and look for outliers that are +/- 4 std. dev. from the mean.

```{r pheno_std_boxplot,fig.height=8}
tmp = pheno %>% 
        select(mouse, pct_wt_loss:serum_cxcl1) %>%  
        pivot_longer(cols = pct_wt_loss:serum_cxcl1, names_to = 'phenotype', values_to = 'value') %>%
        group_by(phenotype) %>%
        mutate(value = log1p(value),
               value = if_else(phenotype == 'lung_necr_ratio', expm1(value), value),
               value = if_else(phenotype == 'lung_necr_ratio', log(value + 1e-4), value),
               value = scale(value))
outliers = tmp %>% 
             filter(abs(value) > 4)
tmp %>%
  ggplot(aes(phenotype, value)) +
    geom_boxplot() +
    geom_hline(aes(yintercept = 4),  color = 'red') +
    geom_hline(aes(yintercept = -4), color = 'red') +
    geom_text_repel(mapping = aes(phenotype, value, label = mouse), data = outliers, color = 'red') +
    coord_flip()
```

Serum Cxc.2 and lung_prob_ss have skewed distributions. 

There are a handfull of outliers.

Look at a violin plot of the data as well. This allows me to see if there are multi-modal distributions that the boxplot hides.

```{r pheno_std_violin,fig.width=8}
pheno %>% 
  select(pct_wt_loss:serum_cxcl1) %>%  
  pivot_longer(cols = everything(), names_to = 'phenotype', values_to = 'value') %>%
  group_by(phenotype) %>%
  mutate(value = log1p(value),
         value = if_else(phenotype == 'lung_necr_ratio', expm1(value), value),
         value = if_else(phenotype == 'lung_necr_ratio', log(value + 1e-4), value),
         value = scale(value)) %>%
  ggplot(aes(phenotype, value)) +
    geom_violin(draw_quantiles = c(0.25, 0.5, 0.75)) +
    geom_hline(aes(yintercept = 4),  color = 'red') +
    geom_hline(aes(yintercept = -4), color = 'red') +
  coord_flip()
```

Some of the distributions are multimodal. This is either good, because they will separate by genotype, or bad, because it indicates different technical effects.

Look at the necrosis percentages by themselves.

```{r necr_hist}
pheno %>% 
  select(mouse, starts_with('lung_necr')) %>% 
  pivot_longer(starts_with('lung_necr'), names_to = 'phenotype', values_to = 'value') %>% 
  ggplot() +
    geom_histogram(aes(value)) +
    scale_x_log10() +
    facet_wrap(~phenotype)
```


# Phenotypes by Mouse ID

Next, I plot each phenotype by sample ID, looking for odd trends over time. I will color the points by Mtb dose.

```{r pheno_time_series,fig.width=12,fig.height=10}
pheno %>% 
  select(mouse, mtb_dose, pct_wt_loss:serum_cxcl1) %>% 
  pivot_longer(cols = pct_wt_loss:serum_cxcl1, names_to = 'phenotype', values_to = 'value') %>%
  mutate(mouse = as.numeric(mouse),
         value = log1p(value)) %>%
  ggplot(aes(mouse, value, color = mtb_dose)) +
    geom_point() +
    scale_color_viridis_c() +
    facet_wrap(~phenotype, scales = 'free_y') +
    labs(title = 'Mouse ID vs. log(phenotype)')
```

Il10 values are decrasing with time. Il12 values for mice > 500 are much lower than before. Vegf values are also lower for mice > 250. Oddly, pct_wt_loss is high in the last batch. Some of these effects my be due to initial Mtb dose. 

# Phenotypes by Dose

Look at the relationship between phenotype values and the Mtb Initital Dose. 

```{r pheno_by_dose,fig.width=10}
pheno %>% 
  select(mouse, mtb_dose, pct_wt_loss:serum_cxcl1) %>% 
  pivot_longer(cols = pct_wt_loss:serum_cxcl1, names_to = 'phenotype', values_to = 'value') %>%
  mutate(value = log1p(value)) %>%
  ggplot(aes(mtb_dose, value, color = mtb_dose)) +
    geom_point(alpha = 0.3) +
    geom_smooth(method = 'lm') +
    facet_wrap(~phenotype) +
    scale_color_viridis_c() +
    labs(title = 'Phenotypes by M.Tb. Dose', y = 'log10(value)') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```

# Phenotypes by Generation

```{r pheno_by_gen,fig.width=10}
pheno %>% 
  select(mouse, gen, pct_wt_loss:serum_cxcl1) %>% 
  pivot_longer(cols = pct_wt_loss:serum_cxcl1, names_to = 'phenotype', values_to = 'value') %>%
  mutate(value = log1p(value)) %>%
  ggplot(aes(gen, value)) +
    geom_boxplot() +
    geom_jitter(alpha = 0.3) +
    facet_wrap(~phenotype) +
    labs(title = 'Phenotypes by Generation', y = 'log10(value)') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```

This looks similar to the results of plotting samples vs phenotype values.

# Phenotypes by Euthanasia Status

Plot each phenotype versus the euthanasia status.

```{r pheno_by_euth,fig.width=10,fig.height=10,message=FALSE}
pheno %>% 
  select(mouse, euth, pct_wt_loss:serum_cxcl1) %>% 
  pivot_longer(cols = pct_wt_loss:serum_cxcl1, names_to = 'phenotype', values_to = 'value') %>%
  mutate(value = log1p(value)) %>%
  left_join(select(pheno, mouse, euth)) %>%
  ggplot(aes(euth, value)) +
    geom_jitter() +
    geom_smooth(method = 'lm') +
    facet_wrap(~phenotype) +
    labs(title = 'Phenotypes by Euthanasia Status')
```

Plot each phenotype versus the euthansia day.

```{r pheno_by_euth_day,fig.width=10,fig.height=10,message=FALSE}
pheno %>% 
  select(mouse, euth_day, pct_wt_loss:serum_cxcl1) %>% 
  pivot_longer(cols = pct_wt_loss:serum_cxcl1, names_to = 'phenotype', values_to = 'value') %>%
  mutate(value = log1p(value)) %>%
  left_join(select(pheno, mouse, euth_day)) %>%
  ggplot(aes(euth_day, value)) +
    geom_jitter() +
    geom_smooth(method = 'lm') +
    facet_wrap(~phenotype) +
    labs(title = 'Phenotypes by Euthanasia Day')
```

# Phenotype ANOVA with covariates.

There may be some influence of batch or Mtb dose on the phenotypes. Note that mach_type is the same for all DO mice. I'm fitting the whole model with all of the covariates. 

```{r pheno_anova}
# I tried to use tidyverse and map() here, but something about the differing number of samples (I think) kept it from working.
tmp = pheno %>% 
         select(mouse, gen, expt, cage, aero_run, mtb_dose, pct_wt_loss:serum_cxcl1) %>%
         mutate(gen      = as.factor(gen),
                cage     = as.factor(cage),
                aero_run = as.factor(aero_run)) %>%
         pivot_longer(cols = c(-mouse, -(gen:mtb_dose)), names_to = 'phenotype', values_to = 'value') %>%
         mutate(value = log1p(value),
                value = if_else(phenotype == 'necr_ratio', expm1(value), value),
                value = if_else(phenotype == 'necr_ratio', log(value + 1e-4), value)) %>%
         pivot_wider(names_from = phenotype, values_from = value)

# P-value results.
pheno_anova = data.frame(phenotype = colnames(tmp)[-(1:6)],
                         n         = 0,
                         pv_gen    = NA,
                         pv_cage   = NA,
                         pv_aero_run = NA,
                         pv_mtb_dose = NA)

# Go through each phenotype and record the ANOVA p-values for the potential covariates.
for(i in 1:nrow(pheno_anova)) {
  
  pheno_name = pheno_anova$phenotype[i]
  
  # Make temporary data.frame.
  df = tmp[,c('gen', 'cage', 'aero_run', 'mtb_dose', pheno_name)]
  df = df[complete.cases(df),]
  colnames(df)[ncol(df)] = 'value'

  # We need enough samples to fit the model.
  if(nrow(df) > 110) {
    # Fit model with all factors.
    mod = lm(value ~ ., data = df)
    av = anova(mod)
    pheno_anova$n[i]           = nrow(df)
    pheno_anova$pv_gen[i]      = av['gen','Pr(>F)']
    pheno_anova$pv_cage[i]     = av['cage','Pr(>F)']
    pheno_anova$pv_aero_run[i] = av['aero_run','Pr(>F)']
    pheno_anova$pv_mtb_dose[i] = av['mtb_dose','Pr(>F)']
  } # if(nrow(df) > 0)

} # for(i)
```

```{r plot_pheno_anova}
pheno_anova %>%
  select(-pv_mtb_dose) %>%
  gather(covar, pvalue, -phenotype, -n) %>%
  mutate(pvalue = -log10(pvalue)) %>%
  ggplot() +
    geom_point(aes(phenotype, pvalue)) +
    facet_wrap(~covar) +
    labs(title = 'ANOVA p-values for phenotype covariates', y = '-log10(p-value)') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```


```{r write_pheno_anova}
write_csv(pheno_anova, file.path(base_dir, 'results', 'phenotype_anova', 'tufts_do_tb_pheno_anova.csv'))

kable(pheno_anova, digits = 3, format.args = list(scientific = TRUE))
```

Generation seems to have the strongest and most consistent effect. Cage is generally not a factor, except for Ifng, Tnf, Il12, Il10 and Vegf. Should we add this in when mapping those traits? Aero_run is also significant much of the time.  Mtb dose dropped out of the model since it is often confounded with generation.

I'm going to fit the same models, but excluding generation to see if initial Mtb dose does a better job of explaining the variance.

```{r pheno_anova_sans_gen}
# P-value results.
pheno_anova = data.frame(phenotype = colnames(tmp)[-(1:6)],
                         n         = 0,
                         pv_cage   = NA,
                         pv_aero_run = NA,
                         pv_mtb_dose = NA)

# Go through each phenotype and record the ANOVA p-values for the potential covariates.
for(i in 1:nrow(pheno_anova)) {
  
  pheno_name = pheno_anova$phenotype[i]
  
  # Make temporary data.frame.
  df = tmp[,c('cage', 'aero_run', 'mtb_dose', pheno_name)]
  df = df[complete.cases(df),]
  colnames(df)[ncol(df)] = 'value'

  # We need enough samples to fit the model.
  if(nrow(df) > 110) {
    # Fit model with all factors.
    mod = lm(value ~ ., data = df)
    av = anova(mod)
    pheno_anova$n[i]           = nrow(df)
    pheno_anova$pv_cage[i]     = av['cage','Pr(>F)']
    pheno_anova$pv_aero_run[i] = av['aero_run','Pr(>F)']
    pheno_anova$pv_mtb_dose[i] = av['mtb_dose','Pr(>F)']
  } # if(nrow(df) > 0)

} # for(i)

kable(pheno_anova, digits = 3, format.args = list(scientific = TRUE))
```

```{r plot_pheno_anova_sans_gen}
pheno_anova %>%
  gather(covar, pvalue, -phenotype, -n) %>%
  mutate(pvalue = -log10(pvalue)) %>%
  ggplot() +
    geom_point(aes(phenotype, pvalue)) +
    facet_wrap(~covar) +
    labs(title = 'ANOVA p-values for phenotype covariates (without Initial Mtb Dose)', y = '-log10(p-value)') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```

I also want to look at how gen by aero_run interactions affects each phenotype.

```{r pheno_gen_aero_interation}
# P-value results.
pheno_anova3 = data.frame(phenotype = colnames(tmp)[-(1:6)],
                          n         = 0,
                          pv_gen   = NA,
                          pv_cage  = NA,
                          pv_aero_run = NA,
                          pv_gen_aero = NA)

# Go through each phenotype and record the ANOVA p-values for the potential covariates.
for(i in 1:nrow(pheno_anova3)) {
  
  pheno_name = pheno_anova3$phenotype[i]
  
  # Make temporary data.frame.
  df = tmp[,c('gen', 'cage', 'aero_run', 'mtb_dose', pheno_name)]
  df = df[complete.cases(df),]
  colnames(df)[ncol(df)] = 'value'

  # We need enough samples to fit the model.
  if(nrow(df) > 110) {
    # Fit model with all factors.
    mod = lm(value ~ gen + cage + aero_run + gen:aero_run, data = df)
    av = anova(mod)
    pheno_anova3$n[i]           = nrow(df)
    pheno_anova3$pv_gen[i]      = av['gen','Pr(>F)']
    pheno_anova3$pv_cage[i]     = av['cage','Pr(>F)']
    pheno_anova3$pv_aero_run[i] = av['aero_run','Pr(>F)']
    pheno_anova3$pv_gen_aero[i] = av['gen:aero_run','Pr(>F)']
  } # if(nrow(df) > 0)

} # for(i)

kable(pheno_anova3, digits = 3, format.args = list(scientific = TRUE))
```

```{r plot_pheno_anova_gen_aero_interaction}
pheno_anova3 %>%
  gather(covar, pvalue, -phenotype, -n) %>%
  mutate(pvalue = -log10(pvalue)) %>%
  ggplot() +
    geom_point(aes(phenotype, pvalue)) +
    facet_wrap(~covar) +
    labs(title = 'ANOVA p-values for phenotype covariates (without gen by aero run interaction)', y = '-log10(p-value)') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```

The generation by aero run term doesn't seem to explain much of the variance.

I still need to think about what to use in the mapping. We can't put cage in the model because it will eat up too many degrees of freedom. It is also re-used, to perhaps it should be gen by cage or expt by cage.

# Bivaraite Distributions

Next, I will plot each pair of phenotypes, colored by generation to see if there are unusual patterns.

```{r pheno_bivariate_dist,fig.width=12,fig.height=12,warning=FALSE,message=FALSE}
pheno %>% 
  select(mouse, gen, pct_wt_loss:serum_cxcl1) %>% 
  pivot_longer(cols = pct_wt_loss:serum_cxcl1, names_to = 'phenotype', values_to = 'value') %>%
  mutate(value = log1p(value)) %>%
  pivot_wider(names_from = phenotype, values_from = value) %>%
  ggpairs(columns = 3:19, ggplot2::aes(color = gen)) +
  theme(panel.spacing = ggplot2::unit(0, 'npc'))
```

Plot the Ccl and Cxcl proteins.

```{r ccl_cscl_pairs,fig.width=10,fig.height=10,message=FALSE}
pheno %>% 
  select(mouse, starts_with('lung_ccl'), starts_with('lung_cxcl')) %>%
  pivot_longer(cols = lung_cxcl5:lung_cxcl1, names_to = 'phenotype', values_to = 'value') %>%
  na.omit() %>%
  mutate(value = log1p(value)) %>%
  pivot_wider(names_from = phenotype, values_from = value) %>%
  left_join(select(pheno, mouse, gen)) %>%
  ggpairs(columns = 2:4, ggplot2::aes(color = gen), alpha = 0.3)
```

It looks like there is something wrong with Cxcl2 in G22. It has tighter variance than the other batches.

Use another view to look at the correlation between phenotypes. This one hides the data, but makes it easier to see which phenotypes are correlated with each other and may warrant further investigations. I can't cluster the phenotypes by similarity because there are NA value in the correlation matrix and most methods blow up. The NA values come from phenotypes that have not been measured in the same mice.

```{r corrplot_pheno,fig.width=12}
corrplot.mixed(cor(as.matrix(pheno[,-(1:17)]), use = 'pair'), lower = 'number', upper = 'ellipse')
```

Plot pct_wt_loss, mtb_burden, cxcl1, cxcl2, cxcl5, mmp8, s100a8 & tnf.

```{r cxcl_more_pairs,fig.width=10,fig.height=10,message=FALSE}
pheno %>%
  select(mouse, gen, pct_wt_loss, mtb_burden, starts_with('lung_cxcl'), lung_mmp8, lung_s100a8, lung_tnf) %>%
  na.omit() %>%
  pivot_longer(cols = pct_wt_loss:lung_tnf, names_to = 'phenotype', values_to = 'value') %>%
  mutate(value = log1p(value)) %>%
  pivot_wider(names_from = phenotype, values_from = value) %>%
  ggpairs(columns = 3:10, ggplot2::aes(color = gen))
```

Plot these colored by whether or not the mouse was euthanized. I may be misinterpreting what the 'euth' column means. It is usually True when the mouse was euthanized and False when it was found dead. Should euthanasia status or lifespan be in the mapping model?

```{r cxcl_more_pairs_color_euth,fig.width=10,fig.height=10,message=FALSE}
pheno %>%
  select(mouse, euth, pct_wt_loss, mtb_burden, starts_with('lung_cxcl'), lung_mmp8, lung_s100a8, lung_tnf) %>%
  na.omit() %>%
  pivot_longer(cols = pct_wt_loss:lung_tnf, names_to = 'phenotype', values_to = 'value') %>%
  mutate(value = log1p(value)) %>%
  pivot_wider(names_from = phenotype, values_from = value) %>%
  ggpairs(columns = 3:10, ggplot2::aes(color = euth))
```

Plot Cxcl2 by dose (confounded wtih generation).

```{r cxcl2_by_dose}
pheno %>%
  select(mouse, mtb_dose, lung_cxcl2) %>%
  na.omit() %>%
  mutate(lung_cxcl2 = log1p(lung_cxcl2)) %>%
  mutate(mtb_dose = factor(mtb_dose, levels = c(0, 15, 16, 20, 28, 35, 97, 127))) %>%
  ggplot(aes(lung_cxcl2, fill = mtb_dose)) +
    geom_density(alpha = 0.3) +
    labs(title = 'Cxcl2')
```

Cxcl2 is often biomodal in these plots. But dose 28 (also G22) looks very skewed to the left. I'm not sure if this is real or an artifact. 

```{r cxcl2_by_gen_violin}
pheno %>%
  select(mouse, lung_cxcl2) %>%
  na.omit() %>%
  mutate(lung_cxcl2 = log1p(lung_cxcl2)) %>%
  left_join(select(pheno, mouse, mtb_dose)) %>%
  mutate(mtb_dose = factor(mtb_dose, levels = c(0, 15, 16, 28, 97, 127))) %>%
  ggplot(aes(mtb_dose, lung_cxcl2)) +
    geom_violin(alpha = 0.3) +
    labs(title = 'Lung Cxcl2')
```

Add columns to pheno for mapping.

```{r add_mapping_columns}
pheno = left_join(pheno, select(orig_pheno, mouse, map_survival, map_lung))
pheno = select(pheno, mouse:euth, map_survival, map_lung, pct_wt_loss:serum_cxcl1)
```


# Write out Phenotypes.

There may be more cleaning or adjustment that needs to be made. For now, I'm going to write out the phenotypes as they are at the end of this script.

```{r write_pheno}
write_csv(pheno, file.path(base_dir, 'data', 'phenotypes', 'tufts_do_tb_pheno_cleaned.csv'))
```

